javascript-review_Moss.js

```diff
@@ -0,0 +1,293 @@
+
+// 1.
+
+// Create an object that has four properties. One property should be set so that the value is a number, another property set to a string, the third to an array, and the fourth to a function. This function should simply console.log the value of the first property.
+
+const obj1 = {
+    name: 'moss',
+    age: 24,
+    hobby: ['game', 'sports'],
+    toString: function() {
+      return this.name + ',' + this.age + ',' + [...this.hobby]
+    }
+  }
+  
+  // 2.
+  
+  // Write a constructor function (to create objects) that takes one argument. It should use that argument to set the value of the first attribute of the object. There should be a second attribute that stores the value ‘red’.
+  
+  // class Obj {
+  //   constructor(first) {
+  //     this.first = first
+  //     this.second = 'red'
+  //   }
+  // }
+  function Obj(first) {
+    return { first: first, second: 'red' }
+  }
+  
+  // Note that a constructor function is the old school syntax that was used in javascript before the sugar class syntax came along.
+  
+  // 3.
+  
+  // Can you access variables defined outside of functions within the scope of functions in JS? Show a simple example of this being used? (Define a constant in global scope and use that in a very simple function.)
+  
+  const NAME = 'Moss'
+  function show() {
+    console.log(NAME)
+  }
+  // show()
+  // 4.
+  
+  // What two ways can be used to access the properties of objects in JS? Give examples of both (although assume the objects have been defined).
+  // 5.
+  // console.log(obj1.name)
+  // console.log(obj1['name'])
+  
+  // What is JSON? Be clear but concise.
+  let q6 =
+    'JSON is javascript object notation is the way javascript display object'
+  // 6.
+  
+  // What is a callback?
+  // 7.
+  let q7 =
+    'Callback are function that is passing as an argument and to be call later on it process of it caller function'
+  
+  // Define a function with that takes two number arguments and a callback. The function will add the two numbers and pass them to the callback as an argument. Now call this function and in the callback simply console.log out the argument that has been passed through.
+  // 8.
+  const time2 = num => {
+    console.log(num)
+    return num * 2
+  }
+  function q8(a, b, time2) {
+    return time2(a + b)
+  }
+  // console.log(q8(1, 2, time2))
+  
+  // Why do we use promises?
+  // 9.
+  const q9 = 'to make javascript run in asychonise way'
+  
+  // What are the two methods at our disposal if we have hold of a promise? What are the names of the functions that are related to these functions (that are called when the promise is fulfilled or fails).
+  // 10.
+  const q10 = '.then and .catch'
+  
+  // What is the DOM?
+  // 11.
+  const q11 =
+    'DOM is like a parent and child directory tree of html element and also refer to how javascript see html page'
```

Comment:
yep that's right, the DOM is also the entire `window` of the browser, not just the html document